src/charm.py

```diff
@@ -152,11 +153,59 @@ def __init__(self, *args) -> None:
     ###################
 
     def _on_livepatch_relation_changed(self, event) -> None:
-        """Handle peer `relation-changed` event."""
-        self._config_changed(event)
+        """
+        On peer relation changed hook.
+
+        This hook is for the non-leader units to get notified when the state
+        changes. On the leader unit this hook should be ignored to avoid
+        repetitive workload restarts while handling relations. This also means,
+        on the leader unit, whenever the state changes, the update workload
+        method should be called manually.
+        """
+        if self.unit.is_leader():
+            return
+        if not self._state.is_ready():
+            logging.warning("State is not ready")
+            return
+        self._update_workload_configuration(event)
+
+    def _restart_non_leader_units(self):
+        """
+        Cue non-leader units to update their workload configuration.
+
+        This method is meant to be used by action hooks that *do not* update the
+        state. It's because upon updating the state, the non-leader units will
+        get receive the cue via the peer-relation-changed event.
```

Comment:
```suggestion
        state. It's because upon updating the state, the non-leader units will
        receive the cue via the peer-relation-changed event.
```